Minor fingerprint improvements

batiati opened the pull request:
- Change the fingerprint for composite keys, taking only the prefix, not the entire key.
Rationale: We don't use composite keys for lookups. However, the bloom filter can be useful for scans with equality conditions by prefix.

- Propagate the key's fingerprint during the prefetch path.
Rationale: Hashing the key only once.

- Change PrefetchKeys' implementation from HashMap to ArrayHashMap.
Rationale: It's better for iteration.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -387,12 +389,16 @@ pub fn GrooveType(
+        };
 
-        const PrefetchKeys = std.AutoHashMapUnmanaged(PrefetchKey, void);
+        const PrefetchKeys = std.ArrayListUnmanaged(PrefetchKey);
```
What is the reason for changing `PrefetchKeys` from a hashset to an arraylist? The former is nice since it automatically deduplicates prefetches. (e.g. imagine that all of the transfers in the batch share the same debit account id -- each one needs to prefetch that same account).

batiati replied:
> The former is nice since it automatically deduplicates prefetches

That's so true, but I suppose the prefetch path now can take care of deduping ids efficiently.

The main reason for using an ArrayList is to remove the inherent cost of hashing the `PrefetchKey` struct and iterating over the hashtable keys, which seems costly compared with the ArrayList.

But I agree it's easier to reason about duplicated ids relying on a hashtable, not the prefetch path.

sentientwaffle replied:
> That's so true, but I suppose the prefetch path now can take care of deduping ids efficiently.

:thinking: Ah, since for e.g. two reads (X and Y) to the same account, either:
- reads X and Y are issued concurrently (and thus coalesced by the grid's read path), or
- read X finishes before Y starts, so Y's lookup_from_memory hits

Is that what you mean?

batiati replied:
Yes, although I haven't measured the worst case (eg a batch using the same two accounts, or even worse one entire lookup batch with the same id for a transfer that doesn't exists).

This approach is too risky, instead I'm benchmarking a ArrayHashMap which can provide both dedup and easy iteration over items.

batiati committed:
Minor fingerprint improvements

Change the fingerprint for composite keys, taking only the prefix, not the entire key.
Rationale: We don't use composite keys for lookups. However, the bloom filter can be useful for scans with equality conditions by prefix.

Propagate the key's fingerprint during the prefetch path.
Rationale: Hashing the key only once.

Change PrefetchKeys' implementation from HashMap to ArrayHashMap.
Rationale: It's better for iteration.

sentientwaffle commented on src/lsm/tree.zig:
```diff
@@ -347,8 +353,12 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type) type {
+            snapshot: u64,
+            key: Key,
+            fingerprint: Fingerprint,
+        ) LookupMemoryResult {
```
Can you add a `if (constants.verify)` check that verifies that `fingerprint` is indeed the fingerprint of the passed `key`? Along with a note that it is provided as an argument (rather than computed here) as an optimization.

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -387,12 +389,16 @@ pub fn GrooveType(
+        };
 
-        const PrefetchKeys = std.AutoHashMapUnmanaged(PrefetchKey, void);
+        const PrefetchKeys = std.AutoArrayHashMapUnmanaged(PrefetchKey, void);
```
Probably wouldn't make much of a perf difference, but if instead of a `(PrefetchKey, void)` hashmap it was a `(union(enum) {id,timestamp}, struct {fingerprint, level}})` hashmap that would slightly reduce the amount of data being hashed.

batiati replied:
I've tried that, not changed, just measurement noise.
Maybe it's a good design for correctness since it should be unique per `union(enum) {id,timestamp}` and not by the whole struct.

batiati committed:
Address review comments

sentientwaffle commented on src/lsm/groove.zig:
```diff
@@ -747,38 +752,49 @@ pub fn GrooveType(
+                switch (prefetch_entry.key_ptr.*) {
                     .id => |id| {
+                        if (constants.verify) {
+                            assert(std.meta.eql(
```
Nit, but I'd suggest `stdx.equal_bytes` to avoid the stack copy.

batiati replied:
Not possible for this type:
```
comptime assert(std.meta.trait.hasUniqueRepresentation(T));
```

sentientwaffle replied:
Ah, nevermind!
